### PR TITLE
[Model][MXNet] neighbor sampling & skip connection & control variate &  graphsage

### PR DESCRIPTION
**Pull request by ZiyueHuang**
## Description
Neighbor sampling for GCN. For the algorithm, you can see Sec. 2.3 in Control Variate paper 
https://arxiv.org/pdf/1710.10568.pdf.

The arg num_hop in current sampler cannot be used, since we need the sampled edges on each hop and on each hop for each node (independently), we need to sample from its whole neighborhood. This sampling procedure should be merged into Sampler in library (the example now is implemented in python, just for discussion and showcase). @zheng-da 

Also we need tutorials about Sampler.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [$CATEGORY] (such as [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [ ] All changes have test coverage
- [ ] Code is well-documented
- [ ] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR

### CODE REVIEW

**Comment by ZiyueHuang** [comment]
Train Control Variate on cora for 3 different trials and the test accuracy always converges to around 80%. @zheng-da 

image

The corresponding log files are here,
[dgl_cv_1.log](https://github.com/dmlc/dgl/files/2753120/dgl_cv_1.log)
[dgl_cv_2.log](https://github.com/dmlc/dgl/files/2753121/dgl_cv_2.log)
[dgl_cv_3.log](https://github.com/dmlc/dgl/files/2753122/dgl_cv_3.log)

**Comment by ZiyueHuang** [comment]
Train Control Variate on cora for 3 different trials and the test accuracy always converges to around 80%.

image

**Comment by ZiyueHuang** [comment]
experiment report

[GCNSampling.pdf](https://github.com/dmlc/dgl/files/2887740/GCNSampling.pdf)

File: examples/mxnet/sampling/gcn_cv_gs_reddit.py [hunk 1]
```diff
@@ -0,0 +1,352 @@
...
+    def forward(self, nf):
+        h = nf.layers[0].data['preprocess']
+        features = nf.layers[0].data['features']
+        h = mx.nd.concat(h, features)
+        h = self.dense1(h)
+        h = self.layer_norm1(h)
+        h = mx.nd.relu(h)
+        h = self.dense2(h)
+        h = self.layer_norm2(h)
+        h = mx.nd.relu(h)
```
Comment (zheng-da) [inline examples/mxnet/sampling/gcn_cv_gs_reddit.py]: The same comments as above

File: examples/mxnet/sampling/gcn_cv_gs_reddit.py [hunk 2]
```diff
@@ -0,0 +1,352 @@
...
+                agg_history_str = 'agg_h_{}'.format(i)
+                g.pull(nf.layer_parent_nid(i+1), fn.copy_src(src='h_{}'.format(i), out='m'),
+                       fn.sum(msg='m', out=agg_history_str))
+
+            node_embed_names = [['preprocess', 'features', 'h_0']]
+            for i in range(1, n_layers):
+                node_embed_names.append(['h_{}'.format(i), 'agg_h_{}'.format(i-1), 'subg_norm', 'norm'])
+            node_embed_names.append(['agg_h_{}'.format(n_layers-1), 'subg_norm', 'norm'])
+
+            nf.copy_from_parent(node_embed_names=node_embed_names)
```
Comment (zheng-da) [inline examples/mxnet/sampling/gcn_cv_gs_reddit.py]: I wonder if we can change the behavior of copy from parent to copy with demand (I.e., copy data from the parent graph when it’s used).

Comment (zheng-da) [reply examples/mxnet/sampling/gcn_cv_gs_reddit.py]: It’s tedious to specify such a complex list

File: examples/mxnet/sampling/gcn_cv_gs_reddit.py [hunk 3]
```diff
@@ -0,0 +1,352 @@
+import argparse, time, math
```
Comment (zheng-da) [inline examples/mxnet/sampling/gcn_cv_gs_reddit.py]: Please rename the file name such as graphsage_cv.py. It’s reddit specific

**Comment by zheng-da** [comment]
Could you move the code of adding self loop out and write a unit test for it? Our PR guidance is doing one thing in a PR. This PR looks very dense.

File: examples/mxnet/sampling/gcn_cv_gs_reddit.py [hunk 4]
```diff
@@ -0,0 +1,352 @@
...
+            pred = infer_model(nf)
+            batch_nids = nf.layer_parent_nid(-1).as_in_context(ctx)
+            batch_labels = labels[batch_nids]
+            num_acc += (pred.argmax(axis=1) == batch_labels).sum().asscalar()
+
+        print("Test Accuracy {:.4f}". format(num_acc/n_test_samples))
+
+
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser(description='GCN')
```
Comment (zheng-da) [inline examples/mxnet/sampling/gcn_cv_gs_reddit.py]: Graphsage

File: examples/mxnet/sampling/gcn_cv_gs_reddit.py [hunk 5]
```diff
@@ -0,0 +1,352 @@
...
+      #Classes %d
+      #Train samples %d
+      #Val samples %d
+      #Test samples %d""" %
+          (n_edges, n_classes,
+              n_train_samples,
+              n_val_samples,
+              n_test_samples))
+
+    # create GCN model
```
Comment (zheng-da) [inline examples/mxnet/sampling/gcn_cv_gs_reddit.py]: Graphsage

File: examples/mxnet/sampling/gcn_cv_gs_reddit.py [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+            h = mx.nd.Dropout(h, p=self.dropout)
+        h = self.dense2(h)
+        if not self.last:
+            h = self.layer_norm2(h)
+            h = mx.nd.relu(h)
+        return {'activation': h}
+
+
+
+class GCNSampling(gluon.Block):
```
[the remark this answers is not in the document]
Comment (ZiyueHuang) [reply examples/mxnet/sampling/gcn_cv_gs_reddit.py]: This is control variate + graphsage. I think `GCNSampling` is fine.

File: examples/mxnet/sampling/gcn_cv_gs_reddit.py [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+        self_h = node.data['self_h']
+        if self.test:
+            h = (h - self_h) * norm
+            # graphsage
+            h = mx.nd.concat(h, self_h)
+        else:
+            agg_history_str = 'agg_h_{}'.format(self.layer_id-1)
+            agg_history = node.data[agg_history_str]
+            subg_norm = node.data['subg_norm']
+            self_delta_h = node.data['self_delta_h']
```
[the remark this answers is not in the document]
Comment (ZiyueHuang) [reply examples/mxnet/sampling/gcn_cv_gs_reddit.py]: How to explain? It's just `h - history` used in control variate.

File: src/graph/sampler.cc [hunk 8]
```diff
@@ -440,6 +441,10 @@ NodeFlow SampleSubgraph(const ImmutableGraph *graph,
                             &tmp_sampled_edge_list,
                             &time_seed);
       }
+      if (add_self_loop) {
+        tmp_sampled_src_list.push_back(dst_id);
+        tmp_sampled_edge_list.push_back(-1);
```
[the remark this answers is not in the document]
Comment (ZiyueHuang) [reply src/graph/sampler.cc]: Yes, but the self loop edges have no corresponding edge data. Users should not use these edge ids if they turn on this flag `add_self_loop`.

File: examples/mxnet/sampling/gcn_cv_sc.py [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+    def forward(self, node):
+        h = node.data['h']
+        if self.test:
+            norm = node.data['norm']
+            h = h * norm
+        else:
+            agg_history_str = 'agg_h_{}'.format(self.layer_id-1)
+            agg_history = node.data[agg_history_str]
+            # control variate
+            h = h + agg_history
```
Comment (zheng-da) [inline examples/mxnet/sampling/gcn_cv_sc.py]: h isn’t normalized?

File: examples/mxnet/sampling/gcn_cv_sc.py [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+            if self.dropout:
+                h = mx.nd.Dropout(h, p=self.dropout)
+        h = self.dense(h)
+        if self.concat:
+            h = mx.nd.concat(h, self.activation(h))
```
Comment (zheng-da) [inline examples/mxnet/sampling/gcn_cv_sc.py]: Is this self loop? It looks different from what you showed me

Comment (zheng-da) [reply examples/mxnet/sampling/gcn_cv_sc.py]: Does its dimension increase along the gcn layers?

File: examples/mxnet/sampling/gcn_cv_sc.py [hunk 11]
```diff
@@ -0,0 +1,321 @@
...
+        for i, layer in enumerate(self.layers):
+            new_history = h.copy().detach()
+            history_str = 'h_{}'.format(i)
+            history = nf.layers[i].data[history_str]
+            h = h - history
+
+            nf.layers[i].data['h'] = h
+            nf.block_compute(i,
+                             fn.copy_src(src='h', out='m'),
+                             lambda node : {'h': node.mailbox['m'].mean(axis=1)},
```
Comment (zheng-da) [inline examples/mxnet/sampling/gcn_cv_sc.py]: Why not use a builtin?

File: examples/mxnet/sampling/gcn_cv_sc.py (hunk 9 above)
Comment (ZiyueHuang) [reply examples/mxnet/sampling/gcn_cv_sc.py]: already normalized in update

File: examples/mxnet/sampling/gcn_cv_sc.py (hunk 10 above)
Comment (ZiyueHuang) [reply examples/mxnet/sampling/gcn_cv_sc.py]: it's skip connection

File: src/graph/sampler.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (zheng-da) [reply src/graph/sampler.cc]: Agree. I wonder if there is a way to tell users not to access edge data if nodefow is created with self loop

File: examples/mxnet/sampling/gcn_cv_sc.py (hunk 10 above)
Comment (ZiyueHuang) [reply examples/mxnet/sampling/gcn_cv_sc.py]: dimension is always 2*hidden

File: examples/mxnet/sampling/gcn_cv_gs_reddit.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (zheng-da) [reply examples/mxnet/sampling/gcn_cv_gs_reddit.py]: Graphsage isn’t gcn?

File: examples/mxnet/sampling/gcn_cv_sc.py (hunk 11 above)
Comment (ZiyueHuang) [reply examples/mxnet/sampling/gcn_cv_sc.py]: no need for manually building a `subg_norm` by using mean

File: examples/mxnet/sampling/gcn_cv_gs_reddit.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (zheng-da) [reply examples/mxnet/sampling/gcn_cv_gs_reddit.py]: I consider graphsage as a different model. Gcn is another model. For example, the graphsage implemented in the other PR calls the model graphsage. I would guess most people consider graphsage as another model

File: src/graph/sampler.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (ZiyueHuang) [reply src/graph/sampler.cc]: I'll add it in the document

File: examples/mxnet/sampling/gcn_cv_gs_reddit.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ZiyueHuang) [reply examples/mxnet/sampling/gcn_cv_gs_reddit.py]: agree

File: examples/mxnet/sampling/gcn_cv_sc.py (hunk 10 above)
Comment (zheng-da) [reply examples/mxnet/sampling/gcn_cv_sc.py]: I mean skip connection. It looks much simpler

File: src/graph/sampler.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (zheng-da) [reply src/graph/sampler.cc]: If it cause segfault, it’s always bad

File: examples/mxnet/sampling/gcn_cv_sc.py (hunk 11 above)
Comment (zheng-da) [reply examples/mxnet/sampling/gcn_cv_sc.py]: Yes, but is the speed ok?

File: examples/mxnet/sampling/gcn_cv_gs_reddit.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (zheng-da) [reply examples/mxnet/sampling/gcn_cv_gs_reddit.py]: i didn't understand what the name `self_delta_h` indicates. my understanding is that this is `h - history` on the nodes with self loops.

File: examples/mxnet/sampling/gcn_cv_gs_reddit.py (hunk 2 above)
Comment (zheng-da) [reply examples/mxnet/sampling/gcn_cv_gs_reddit.py]: this comment is for myself.

File: src/graph/sampler.cc (hunk 8 above)
[the remark this answers is not in the document]
Comment (zheng-da) [reply src/graph/sampler.cc]: I just realized a use case that might have problems with this implementation. In any heterogeneous graph, edge type is stored as edge data. After we sample a NodeFlow from such a graph, we have to use edge id to get the edge data. Since heterogeneous graphs are common graph data, we have to deal with this problem very carefully.

[the remark this answers is not in the document]
Comment (zheng-da) [reply src/graph/sampler.cc]: could you add a check in https://github.com/dmlc/dgl/pull/322/files#diff-5bf42cb930a666d2dc57d12edcb83d59L357?

**Comment by zheng-da** [comment]
everything else looks good.

**Comment by hbsun2113** [comment]
Do you have a plan of neighbor sampling on Pytorch?

**Comment by zheng-da** [comment]
@hbsun2113 the neighbor sampler works for pytorch. would you like to contribute and implement the model in pytorch?

**Comment by hbsun2113** [comment]
> @hbsun2113 the neighbor sampler works for pytorch. would you like to contribute and implement the model in pytorch?

Are there some model examples making use of the sampler on pytorch?  Could you give me a link? I have no idea about how to using neighbor sampler.

**Comment by hbsun2113** [comment]
> the neighbor sampler works for pytorch. would you like to contribute and implement the model in pytorch?

@zheng-da Are there some model examples making use of the sampler on pytorch? Could you give me a link? I have no idea about how to using neighbor sampler.

**Comment by zheng-da** [comment]
@hbsun2113 we don't have an example so far. but you can use it in exactly the same way as https://github.com/dmlc/dgl/pull/322/files#diff-21ce5c875a30f73b451393838c3cbf49R231
